File: Nenakhov/2192.py
class Solution(object):
    def __init__(self):
        self.childs = dict({})
        self.ancestors = []
    def dfs(self, ancestor, node):
        if node in self.childs:
            for child in self.childs[node]: #DFS-ом будем передавать предка 
                if ancestor not in self.ancestors[child]:
                    self.ancestors[child].append(ancestor)
                    self.dfs(ancestor, child)
    
    
    def getAncestors(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[List[int]]
        """
        # будем для каждого родителя хранить детей
        for parent, child in edges:
            if parent in self.childs:
                self.childs[parent].append(child)
            else:
                self.childs[parent] = []
                self.childs[parent].append(child)

        self.ancestors = [[] for i in range(n)]

        for i in range(n):
            self.dfs(i,i)
        
        return self.ancestors

File: Ивашечкин/2192.go
func getAncestors(n int, edges [][]int) [][]int {
    answer := make ([][]int, n)
    matrix := make([][]int, n)
    
    //для адекватного вида матрицы смежности, т.к. даны лишь рёбра 
    for _, i := range edges {
        matrix[i[0]] = append(matrix[i[0]], i[1])
    }


    for i:=0; i<n; i++ {
        bfs(i, matrix, answer)
    }
    return answer
}

func bfs(i int, edges [][]int, answer [][]int)  {
    visited := make(map[int]bool)
    visited[i] = true

    nextNodes := make([]int, 0)
    nextNodes = append(nextNodes, i)
    for len(nextNodes) > 0 {
        node := nextNodes[0]
        for _, j := range edges[node] {
            if visited[j] == false {
                visited[j] = true
                nextNodes = append(nextNodes, j)
                answer[j] = append(answer[j], i)
            }
        }
        nextNodes = nextNodes[1:]
    }
}